README.md
# tap-wonolo
[![PyPI version](https://badge.fury.io/py/tap-wonolo.svg)](https://badge.fury.io/py/tap-wonolo)
![PyPI - Status](https://img.shields.io/pypi/status/tap-wonolo)
[![Build Status](https://travis-ci.com/goodeggs/tap-wonolo.svg?branch=master)](https://travis-ci.com/goodeggs/tap-wonolo.svg?branch=master)
![PyPI - Python Version](https://img.shields.io/pypi/pyversions/tap-wonolo)
[![License: GPL v3](https://img.shields.io/badge/License-GPLv3-blue.svg)](https://www.gnu.org/licenses/gpl-3.0)

A [Singer](https://www.singer.io/) tap for extracting data from the [Wonolo REST API v2](https://wonolo.readme.io/docs/getting-started).

## Installation

Since package dependencies tend to conflict between various taps and targets, Singer [recommends](https://github.com/singer-io/getting-started/blob/master/docs/RUNNING_AND_DEVELOPING.md#running-singer-with-python) installing taps and targets into their own isolated virtual environments:

### Install Wonolo Tap

```bash
$ python3 -m venv ~/.venvs/tap-wonolo
$ source ~/.venvs/tap-wonolo/bin/activate
$ pip3 install tap-wonolo
$ deactivate
```

### Install Stitch Target (optional)

```bash
$ python3 -m venv ~/.venvs/target-stitch
$ source ~/.venvs/target-stitch/bin/activate
$ pip3 install target-stitch
$ deactivate
```

## Configuration

The tap accepts a JSON-formatted configuration file as arguments. This configuration file has four required fields:

1. `api_key`: A valid Wonolo API key.
2. `secret_key`: A valid Wonolo secret key.
3. `environment`: A valid Wonolo enviroment (either "test" or "production").

An bare-bones Wonolo configuration may file may look like the following:

```json
{
  "api_key": "foo",
  "secret_key": "bar",
  "environment": "test"
}
```

### Authentication
If no `auth_token` key is supplied in the configuration file, the tap will automatically request one via the API and write it back to the config file:

```json
{
  "api_key": "foo",
  "secret_key": "bar",
  "environment": "test",
  "auth_token": "foobar",
  "auth_token_expires_at": "2019-11-08T00:00:20Z",
}
```

The tap will then use the aforementioned `auth_token` to authenticate to the API, until the `auth_token` becomes invalidated, at which point the tap will automatically request and record a new `auth_token` from the API.

### Granular Stream Configuration

Additionally, you may specify more granular configurations for individual streams. Each key under a stream should represent a valid API request parameter for that endpoint. A more fleshed-out configuration file may look similar to the following:

```json
{
  "api_key": "foo",
  "secret_key": "bar",
  "environment": "test",
  "api_version": "v2",
  "streams": {
    "jobs": {
      "state": "fulfilled",
      "company_id": "1234",
      "classification": "w2"
    },
    "job_requests": {
      "w2_hourly_rate": 21.5,
      "updated_after": "2017-05-24T17:01:19.391-07:00"
    }
  }
}
```

## Streams

The current version of the tap syncs three distinct [Streams](https://github.com/singer-io/getting-started/blob/master/docs/SYNC_MODE.md#streams):
1. `Jobs`: [Endpoint Documentation](https://wonolo.readme.io/docs/entities-in-the-api#section-jobs)
2. `Job Requests`: [Endpoint Documentation](https://wonolo.readme.io/docs/entities-in-the-api#section-job-requests)
3. `Users`: [Endpoint Documentation](https://wonolo.readme.io/docs/entities-in-the-api#section-users)

## Discovery

Singer taps describe the data that a stream supports via a [Discovery](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#discovery-mode) process. You can run the Dayforce tap in Discovery mode by passing the `--discover` flag at runtime:

```bash
$ ~/.venvs/tap-wonolo/bin/tap-wonolo --config=config/wonolo.config.json --discover
```

The tap will generate a [Catalog](https://github.com/singer-io/getting-started/blob/master/docs/DISCOVERY_MODE.md#the-catalog) to stdout. To pass the Catalog to a file instead, simply redirect it to a file:s

```bash
$ ~/.venvs/tap-wonolo/bin/tap-wonolo --config=config/wonolo.config.json --discover > catalog.json
```

## Sync to stdout

Running a tap in [Sync mode](https://github.com/singer-io/getting-started/blob/master/docs/SYNC_MODE.md#sync-mode) will extract data from the various selected Streams. In order to run a tap in Sync mode and have messages emitted to stdout, pass a valid configuration file and catalog file:

```bash
$ ~/.venvs/tap-wonolo/bin/tap-wonolo --config=config/wonolo.config.json --catalog=catalog.json
```

The tap will emit occasional [Metric](https://github.com/singer-io/getting-started/blob/master/docs/SYNC_MODE.md#metric-messages), [Schema](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#schema-message), [Record](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#record-message), and [State messages](https://github.com/singer-io/getting-started/blob/master/docs/SPEC.md#state-message). You can persist State between runs by redirecting messages to a file:

```bash
$ ~/.venvs/tap-wonolo/bin/tap-wonolo --config=config/wonolo.config.json --catalog=catalog.json >> state.json
$ tail -1 state.json > state.json.tmp
$ mv state.json.tmp state.json
```

## Sync to Stitch

You can also send the output of the tap to [Stitch Data](https://www.stitchdata.com/) for loading into the data warehouse. To do this, first create a JSON-formatted configuration for Stitch. This configuration file has two required fields:
1. `client_id`: The ID associated with the Stitch Data account you'll be sending data to.
2. `token` The token associated with the specific [Import API integration](https://www.stitchdata.com/docs/integrations/import-api/) within the Stitch Data account.

An example configuration file will look as follows:

```json
{
  "client_id": 1234,
  "token": "foobarfoobar"
}
```

Once the configuration file is created, simply pipe the output of the tap to the Stitch Data target and supply the target with the newly created configuration file:

```bash
$ ~/.venvs/tap-wonolo/bin/tap-wonolo --config=config/dayforce.config.json --catalog=catalog.json --state=state.json | ~/.venvs/target-stitch/bin/target-stitch --config=config/stitch.config.json >> state.json
$ tail -1 state.json > state.json.tmp
$ mv state.json.tmp state.json
```

## Contributing

The first step to contributing is getting a copy of the source code. First, [fork `tap-wonolo` on GitHub](https://github.com/goodeggs/tap-wonolo/fork). Then, `cd` into the directory where you want your copy of the source code to live and clone the source code:

```bash
$ git clone git@github.com:YourGitHubName/tap-wonolo.git
```

Now that you have a copy of the source code on your local machine, you can leverage [Pipenv](https://docs.pipenv.org/en/latest/) and the corresponding `Pipfile` to install of the development dependencies within a virtual environment:

```bash
$ pipenv install --three --dev
```

This command will create an isolated virtual environment for your `tap-wonolo` project and install all the development dependencies defined within the `Pipfile` inside of the environment. You can then enter a shell within the environment:

```bash
$ pipenv shell
```

Or, you can run individual commands within the environment without entering the shell:

```bash
$ pipenv run <command>
```

For example, to format your code using [isort](https://github.com/timothycrosley/isort) and [flake8](http://flake8.pycqa.org/en/latest/index.html) before commiting changes, run the following commands:

```bash
$ pipenv run make isort
$ pipenv run make flake8
```

You can also run the entire testing suite before committing using [tox](https://tox.readthedocs.io/en/latest/):

```bash
$ pipenv run tox
```

Finally, you can run your local version of the tap within the virtual environment using a command like the following:

```bash
$ pipenv run tap-wonolo --config=config/dayforce.config.json --catalog=catalog.json
```

Once you've confirmed that your changes work and the testing suite passes, feel free to put out a PR!


CHANGELOG.md
## 0.1.2

 - Updated `JobRequests` schema to accurately reflect the schema of the `badge_requirements` field.

## 0.1.1

 - Added fibonaccial decay backoff to the `_get()` method on the `WonoloStream`.
 - Made Rollbar logging more explicitly optional.
 - Loosened restrictions on required package versions.


tests/conftest.py
import argparse
import json

import pytest

from tap_wonolo.client import JobRequestsStream, JobsStream, UsersStream


@pytest.fixture(scope='function')
def config(shared_datadir):
    with open(shared_datadir / 'test.config.json') as f:
        return json.load(f)


@pytest.fixture(scope='function')
def state(shared_datadir):
    with open(shared_datadir / 'test.state.json') as f:
        return json.load(f)


@pytest.fixture(scope='function')
def args(config, state, shared_datadir):
    args = argparse.Namespace()
    setattr(args, 'config', config)
    setattr(args, 'state', state)
    setattr(args, 'config_path', shared_datadir / 'test.config.json')
    setattr(args, 'state_path', shared_datadir / 'test.state.json')
    return args


@pytest.fixture(scope='function', params={JobsStream, JobRequestsStream, UsersStream})
def client(config, state, shared_datadir, request):
    return request.param(api_key=config.get("api_key"),
                         secret_key=config.get("secret_key"),
                         api_version=config.get("api_version"),
                         environment=config.get("environment"),
                         config=config,
                         config_path=shared_datadir / 'test.config.json',
                         state=state)


tests/test_client.py
import pytest
import requests
import responses
from singer.schema import Schema

from tap_wonolo.client import is_fatal_code


@pytest.mark.parametrize('status_code', [400, 401, 403, 404,
                                         pytest.param(500, marks=pytest.mark.xfail),
                                         pytest.param(502, marks=pytest.mark.xfail),
                                         pytest.param(503, marks=pytest.mark.xfail),
                                         pytest.param(504, marks=pytest.mark.xfail)])
def test_is_fatal_code(status_code):
    resp = requests.models.Response()
    resp.status_code = status_code
    exc = requests.exceptions.RequestException(response=resp)
    assert is_fatal_code(exc)


def test_client_from_args_class_method(client, args, config, state, shared_datadir):
    stream = client.from_args(args)
    assert stream.api_key == "foo"
    assert stream.secret_key == "bar"
    assert stream.api_version == "v2"
    assert stream.environment == "test"
    assert stream.config == config
    assert stream.state == state
    assert stream.config_path == shared_datadir / 'test.config.json'


def test_load_schema(client):
    schema = client._load_schema()
    assert isinstance(schema, dict)
    assert Schema.from_dict(schema)


def test_get_successful(client):
    with responses.RequestsMock() as rsps:
        expected = {client.tap_stream_id: [{"id": 1234}, {"id": 2345}]}
        rsps.add(responses.GET, f"{client.base_url}/{client.tap_stream_id}",
                 json=expected, status=200)
        resp = client._get(endpoint=f"/{client.tap_stream_id}")
        assert resp == expected


@pytest.mark.xfail()
@pytest.mark.parametrize('status_code', [400, 401, 403, 404])
def test_get_fatal(client, status_code):
    with responses.RequestsMock() as rsps:
        expected = {}
        rsps.add(responses.GET, f"{client.base_url}/{client.tap_stream_id}",
                 json=expected, status=status_code)
        resp = client._get(endpoint=f"/{client.tap_stream_id}")
        assert resp == expected


tap_wonolo/__init__.py
import json
import os
import sys

import rollbar
import singer

from .client import JobRequestsStream, JobsStream, UsersStream

AVAILABLE_STREAMS = {
    JobsStream,
    JobRequestsStream,
    UsersStream
}

REQUIRED_CONFIG_KEYS = [
    "api_key",
    "secret_key",
    "environment"
]

LOGGER = singer.get_logger()

try:
    ROLLBAR_ACCESS_TOKEN = os.environ["ROLLBAR_ACCESS_TOKEN"]
    ROLLBAR_ENVIRONMENT = os.environ["ROLLBAR_ENVIRONMENT"]
except KeyError:
    LOGGER.info("No Rollbar environment variables found. Rollbar logging disabled..")
    log_to_rollbar = False
else:
    rollbar.init(ROLLBAR_ACCESS_TOKEN, ROLLBAR_ENVIRONMENT)
    log_to_rollbar = True


def discover(args, select_all=False):
    LOGGER.info('Starting discovery..')

    catalog = {"streams": []}
    for available_stream in AVAILABLE_STREAMS:
        stream = available_stream.from_args(args)
        catalog_entry = {
            "stream": stream.tap_stream_id,
            "tap_stream_id": stream.tap_stream_id,
            "schema": stream.schema,
            "metadata": singer.metadata.get_standard_metadata(schema=stream.schema,
                                                              key_properties=stream.key_properties,
                                                              valid_replication_keys=stream.bookmark_properties,
                                                              replication_method=stream.replication_method)
        }

        if select_all is True:
            catalog_entry["metadata"][0]["metadata"]["selected"] = True
        catalog["streams"].append(catalog_entry)

    print(json.dumps(catalog, indent=2))


def sync(args):
    LOGGER.info('Starting sync..')

    if not args.state:
        args.state = {"bookmarks": {}}

    selected_streams = {catalog_entry.stream for catalog_entry in args.catalog.get_selected_streams(args.state)}

    for available_stream in AVAILABLE_STREAMS:
        if available_stream.tap_stream_id in selected_streams:
            stream = available_stream.from_args(args)
            LOGGER.info(f"Starting sync for stream {stream.tap_stream_id}..")
            singer.bookmarks.set_currently_syncing(state=args.state, tap_stream_id=stream.tap_stream_id)
            stream.write_state_message()
            stream.write_schema_message()
            stream.sync()
            singer.bookmarks.set_currently_syncing(state=args.state, tap_stream_id=None)
            stream.write_state_message()


def _main():
    args = singer.parse_args(required_config_keys=REQUIRED_CONFIG_KEYS)
    if args.discover:
        discover(args, select_all=True)
    else:
        sync(args)


def main():
    try:
        _main()
    except Exception:
        if log_to_rollbar is True:
            LOGGER.info("Reporting exception info to Rollbar..")
            rollbar.report_exc_info()
        LOGGER.exception(msg="Uncaught Exception..")
        sys.exit(1)


if __name__ == "__main__":
    main()


tap_wonolo/client.py
import json
import os
from datetime import datetime
from typing import Dict, Generator, List, Optional, Set

import attr
import backoff
import requests
import singer

from .version import __version__

LOGGER = singer.get_logger()


def is_fatal_code(e: requests.exceptions.RequestException) -> bool:
    '''Helper function to determine if a Requests reponse status code
    is a "fatal" status code. If it is, the backoff decorator will giveup
    instead of attemtping to backoff.'''
    return 400 <= e.response.status_code < 500 and e.response.status_code != 429


@attr.s
class WonoloStream(object):
    tap_stream_id: Optional[str] = None

    api_key: str = attr.ib()
    secret_key: str = attr.ib()
    environment: str = attr.ib(validator=attr.validators.in_(["test", "production"]))
    config: Dict = attr.ib(repr=False)
    config_path: str = attr.ib()
    state: Dict = attr.ib()
    base_url: str = attr.ib(init=False)
    schema: Dict = attr.ib(init=False)
    auth_token: Optional[str] = attr.ib(repr=False, default=None)
    auth_token_expires_at: Optional[str] = attr.ib(default=None)
    api_version: str = attr.ib(default="v2", validator=attr.validators.instance_of(str))
    params: Dict = attr.ib(init=False, default=None)

    def __attrs_post_init__(self):
        if self.environment == "test":
            self.base_url = f"https://test.wonolo.com/api_{self.api_version}"
        elif self.environment == "production":
            self.base_url = f"https://api.wonolo.com/api_{self.api_version}"

        if self.tap_stream_id is not None:
            self.schema = self._load_schema()

        if self.config.get("streams") is None:
            self.params = {}
        else:
            self.params = self.config.get("streams", {}).get(self.tap_stream_id, {})
            if not isinstance(self.params, dict):
                raise TypeError("Stream parameters must be supplied as JSON.")
            else:
                for key in self.params.keys():
                    if key not in self.valid_params:
                        raise ValueError(f"{key} is not a valid parameter for stream {self.tap_stream_id}")

    @classmethod
    def from_args(cls, args):
        return cls(api_key=args.config.get("api_key"),
                   secret_key=args.config.get("secret_key"),
                   auth_token=args.config.get("auth_token"),
                   auth_token_expires_at=args.config.get("auth_token_expires_at"),
                   api_version=args.config.get("api_version"),
                   environment=args.config.get("environment"),
                   config=args.config,
                   config_path=args.config_path,
                   state=args.state)

    def _get_abs_path(self, path: str) -> str:
        return os.path.join(os.path.dirname(os.path.realpath(__file__)), path)

    def _load_schema(self) -> Dict:
        '''Loads a JSON schema file for a given
        Dayforce resource into a dict representation.
        '''
        schema_path = self._get_abs_path("schemas")
        return singer.utils.load_json(f"{schema_path}/{self.tap_stream_id}.json")

    def _construct_headers(self) -> Dict:
        '''Constructs a standard set of headers for HTTPS requests.'''
        headers = requests.utils.default_headers()
        headers["User-Agent"] = f"python-wonolo-tap/{__version__}"
        headers["Content-Type"] = "application/x-www-form-urlencoded"
        headers["Cache-Control"] = "no-cache"
        headers["Content-Length"] = "0"
        headers["Date"] = singer.utils.strftime(singer.utils.now(), '%a, %d %b %Y %H:%M:%S %Z')
        return headers

    @backoff.on_exception(backoff.fibo,
                          requests.exceptions.HTTPError,
                          max_time=120,
                          giveup=is_fatal_code,
                          logger=LOGGER)
    @backoff.on_exception(backoff.fibo,
                          (requests.exceptions.ConnectionError,
                           requests.exceptions.Timeout),
                          max_time=120,
                          logger=LOGGER)
    def _get(self, endpoint: str, params: Dict = None) -> Dict:
        '''Constructs a standard way of making
        a GET request to the Wonolo REST API.
        '''
        url = self.base_url + endpoint
        headers = self._construct_headers()
        response = requests.get(url, headers=headers, params=params)
        response.raise_for_status()
        return response.json()

    def _post(self, endpoint: str, params: Dict = None, data: Dict = None) -> Dict:
        '''Constructs a standard way of making
        a POST request to the Wonolo REST API.
        '''
        url = self.base_url + endpoint
        headers = self._construct_headers()
        response = requests.post(url, headers=headers, params=params, data=data)
        response.raise_for_status()
        return response.json()

    def _generate_auth_token(self) -> Dict:
        '''Uses the API Key and Secret Key provided in the config file
        to generate an Authorization Token using the /authenticate
        API endpoint.
        '''
        data = {
            "api_key": self.api_key,
            "secret_key": self.secret_key
        }
        return self._post(endpoint='/authenticate', data=data)

    def _save_auth_token(self, auth_response: Dict):
        '''Uses the response provided from the /authenticate endpoint
        to save a new auth token and auth token expiration date
        to the exitsing config file.
        '''
        self.auth_token = auth_response.get("token")
        self.auth_token_expires_at = auth_response.get("expires_at")
        self.config["auth_token"] = auth_response.get("token")
        self.config["auth_token_expires_at"] = auth_response.get("expires_at")
        LOGGER.info('Generating new config..')
        with open(self.config_path, 'r+') as f:
            json.dump(self.config, f, indent=2)

    def _check_auth_token(self):
        '''Checks the provided config for a valid auth token.'''
        if self.auth_token is None:
            LOGGER.info("Generating new auth token..")
            response = self._generate_auth_token()
            self._save_auth_token(response)
        elif datetime.strptime(self.auth_token_expires_at, '%Y-%m-%dT%H:%M:%SZ') <= datetime.utcnow():
            LOGGER.info("Generating new auth token..")
            response = self._generate_auth_token()
            self._save_auth_token(response)
        else:
            LOGGER.info("Using existing auth token..")

    def _yield_records(self, entity: str, params: Optional[Dict] = None) -> Generator[Dict, None, None]:
        '''Yeild individual records for a given entity.'''
        self._check_auth_token()
        if params is None:
            params = {}
        params.update({
            "token": self.auth_token,
            "page": 1,
            "per": 50
        })
        records = self._get(endpoint=f"/{entity}", params=params).get(entity)
        for record in records:
            yield record

        # Paginate
        while len(records) == params.get("per"):
            params["page"] += 1
            records = self._get(endpoint=f"/{entity}", params=params).get(entity)
            if len(records) > 0:
                for record in records:
                    yield record

    def sync(self):
        '''Sync data according to Singer spec.'''
        current_bookmark_str = singer.bookmarks.get_bookmark(state=self.state,
                                                             tap_stream_id=self.tap_stream_id,
                                                             key=self.bookmark_properties)

        if current_bookmark_str is not None:
            self.params.update({self.api_bookmark_param: current_bookmark_str})
            current_bookmark_dt = singer.utils.strptime_to_utc(current_bookmark_str)
        else:
            current_bookmark_dt = None

        with singer.metrics.job_timer(job_type=f"sync_{self.tap_stream_id}"):
            with singer.metrics.record_counter(endpoint=self.tap_stream_id) as counter:
                for record in self._yield_records(entity=self.tap_stream_id, params=self.params):
                    record_bookmark_dt = singer.utils.strptime_to_utc(record.get(self.bookmark_properties))
                    if current_bookmark_dt is None or (record_bookmark_dt > current_bookmark_dt):
                        singer.bookmarks.write_bookmark(state=self.state,
                                                        tap_stream_id=self.tap_stream_id,
                                                        key=self.bookmark_properties,
                                                        val=record.get(self.bookmark_properties))
                        current_bookmark_dt = record_bookmark_dt

                    with singer.Transformer() as transformer:
                        transformed_record = transformer.transform(data=record, schema=self.schema)
                        singer.write_record(stream_name=self.tap_stream_id, time_extracted=singer.utils.now(), record=transformed_record)
                        counter.increment()

    def write_schema_message(self):
        '''Writes a Singer schema message.'''
        return singer.write_schema(stream_name=self.tap_stream_id, schema=self.schema, key_properties=self.key_properties)

    def write_state_message(self):
        '''Writes a Singer state message.'''
        return singer.write_state(self.state)


@attr.s
class JobsStream(WonoloStream):
    tap_stream_id: str = 'jobs'
    key_properties: List[str] = ["id"]
    bookmark_properties: str = "updated_at"
    api_bookmark_param: str = "updated_after"
    replication_method: str = 'INCREMENTAL'
    valid_params: Set[str] = {
        "state",
        "job_request_id",
        "classification",
        "w2_hourly_rate",
        "w2_pay_status",
        "updated_before",
        "updated_after"
    }


@attr.s
class JobRequestsStream(WonoloStream):
    tap_stream_id: str = 'job_requests'
    key_properties: List[str] = ["id"]
    bookmark_properties: str = "updated_at"
    api_bookmark_param: str = "updated_after"
    replication_method: str = 'INCREMENTAL'
    valid_params: Set[str] = {
        "state",
        "company_id",
        "multi_day_job_request_id",
        "classification",
        "w2_hourly_rate",
        "updated_before",
        "updated_after",
        "agent_code"
    }


@attr.s
class UsersStream(WonoloStream):
    tap_stream_id: str = 'users'
    key_properties: List[str] = ["id"]
    bookmark_properties: str = "updated_at"
    api_bookmark_param: str = "updated_after"
    replication_method: str = 'INCREMENTAL'
    valid_params: Set[str] = {
        "type",
        "email",
        "first_name",
        "last_name",
        "external_id",
        "onboarding_last_state",
        "w2_onboarding_status",
        "w2_employee_id",
        "address_state",
        "drug_tested",
        "updated_before",
        "updated_after"
    }
